lib/pkcs11h-core.c

```diff
@@ -1345,3 +1656,97 @@ __pkcs11h_forkFixup () {
 
 #endif				/* !WIN32 */
 
+static
+_pkcs11h_provider_t
+__pkcs11h_get_pkcs11_provider(const char * const reference) {
+	_pkcs11h_provider_t provider = NULL;
+
+	provider = _g_pkcs11h_data->providers;
+	while (
+		provider != NULL &&
+		strcmp (reference, provider->reference)
+	) {
+		provider = provider->next;
+	}
+
+	return provider;
+}
+
+static
+const char*
+__pkcs11h_property_type_to_str(const pkcs11h_provider_property_type_t property) {
+	switch (property) {
+		case PKCS11H_PROVIDER_PROPERTY_LOCATION:
+			return "PKCS11H_PROVIDER_PROPERTY_LOCATION";
+		case PKCS11H_PROVIDER_PROPERTY_ALLOW_PROTECTED_AUTH:
+			return "PKCS11H_PROVIDER_PROPERTY_ALLOW_PROTECTED_AUTH";
+		case PKCS11H_PROVIDER_PROPERTY_MASK_PRIVATE_MODE:
+			return "PKCS11H_PROVIDER_PROPERTY_MASK_PRIVATE_MODE";
+		case PKCS11H_PROVIDER_PROPERTY_SLOT_EVENT_METHOD:
+			return "PKCS11H_PROVIDER_PROPERTY_SLOT_EVENT_METHOD";
+		case PKCS11H_PROVIDER_PROPERTY_SLOT_POLL_INTERVAL:
+			return "PKCS11H_PROVIDER_PROPERTY_SLOT_POLL_INTERVAL";
+		case PKCS11H_PROVIDER_PROPERTY_CERT_IS_PRIVATE:
+			return "PKCS11H_PROVIDER_PROPERTY_CERT_IS_PRIVATE";
+		case PKCS11H_PROVIDER_PROPERTY_INIT_ARGS:
+			return "PKCS11H_PROVIDER_PROPERTY_INIT_ARGS";
+		default:
+			_PKCS11H_ASSERT (FALSE);
+	}
+}
+
+static
+CK_RV
+__pkcs11h_property_value_to_str(const pkcs11h_provider_property_type_t property, const pkcs11h_provider_property_t value, char * * const value_str) {
```

Comment:
oh, now I see, you have nested switch here, probably a better idea is to move this logic to the switch of the set property and have a single logic per property